// facebook-tools/src/getVideoUrl.js
const request = require("request-promise");
const cheerio = require("cheerio");

/**
 * Get raw video URL from a post
 * @param {string} url - URL of the video
 * @returns {string} raw video url
 */
async function getVideoUrl(url) {
	if (typeof url !== "string") throw new Error("URL must be string");

	url = new URL(url);

	let next = false;

	// Validation URL
	switch (url.hostname) {
	case "www.facebook.com":
	case "facebook.com":
	case "m.facebook.com":
	case "mbasic.facebook.com":
	case "fb.com":
		next = true;
		break;
	default:
		next = false;
		break;
	}

	if (!next) throw new Error("Invalid URL!");
	if (url.protocol !== "https:" && url.protocol !== "http:") throw new Error("Invalid protocol");
	if (url.protocol !== "http:" && url.protocol !== "https:") throw new Error("Invalid protocol");

	let options = {
		url: "https://fbdown.net/download.php",
		formData: {
			URLz: url.toString()
		}
	};
    
	let response;

	try {
		response = await request.post(options);
	} catch (e) {
		throw new Error("ERR: Error when trying to request or maybe link is invalid");
	}

	let $;

	try {
		$ = cheerio.load(response);
	} catch (e) {
		throw new Error("ERR: Error when trying to load response");
	}
    
	let obj = {
		sdLink: $("#sdlink").attr("href"),
		hdLink: $("#hdlink").attr("href")
	};

	return obj;
}

module.exports = getVideoUrl;
